capstone/src/main/java/com/google/sps/servlets/UpdateAssociationServlet.java

```diff
@@ -12,23 +18,56 @@
 @WebServlet("/update-associations")
 public class UpdateAssociationServlet extends HttpServlet {
 
+  public static final String SURVEY_ENTITY_KIND = "Actual";
+  public static final String COMMENT_PROPERTY = "text";
+  public static final String RESULT_ENTITY_KIND = "AssociationResult";
+
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
     response.setContentType("text/html;");
     AssociationAnalysis association = new AssociationAnalysis();
     CloudNLPAssociation nlp = new CloudNLPAssociation(LanguageServiceClient.create());
     ArrayList<AssociationResult> res =
-        association.calculateScores(nlp.analyzeAssociations(getComments()));
+        association.calculateScores(nlp.analyzeAssociations(getComments(datastore)));
     response.getWriter().println(res);
+
+    clearPreviousResults(datastore);
+    storeResults(datastore, res);
+
     nlp.close();
   }
 
-  private ArrayList<String> getComments() {
-    return new ArrayList(
-        Arrays.asList(
-            "the police keep on pulling me over and it frustrates me",
-            "I don't understand why the police have to be so militarized",
-            "funding should be redistributed to other social services",
-            "i haven't had any bad experiences with the police personally"));
+  private ArrayList<String> getComments(DatastoreService datastore) {
+    Query query = new Query(SURVEY_ENTITY_KIND);
+    PreparedQuery results = datastore.prepare(query);
+    
+    ArrayList<String> comments = new ArrayList<String>();
+    for (Entity e : results.asIterable()) {
+      comments.add((String) e.getProperty(COMMENT_PROPERTY));
+    }
+    return comments;
+  }
+
+  private void clearPreviousResults(DatastoreService datastore) {
+    Query query = new Query(RESULT_ENTITY_KIND);
+    PreparedQuery results = datastore.prepare(query);
+    
+    ArrayList<Key> toDelete = new ArrayList<Key>();
+    for (Entity entity : results.asIterable()) {
+      toDelete.add(entity.getKey());
+    }
+
+    datastore.delete(toDelete);
+  }
+
+  private void storeResults(DatastoreService datastore, ArrayList<AssociationResult> res) {
+    for (AssociationResult association : res) {
+      Entity entity = new Entity(RESULT_ENTITY_KIND);
+      entity.setProperty("name", association.getContent());
+      entity.setProperty("score", association.getScore());
+      datastore.put(entity);
```

Comment:
When putting new entities on data store it reduces overhead by only making one service call (i.e. call datastore.put(entitiy) once). You can put an array of entities into data store to achieve that result.

https://cloud.google.com/appengine/docs/standard/java/datastore/entities#Batch_operations

Reply:
Changed this to a batch operation!